### PR TITLE
LEST-71 - LuaLS annotations

### PR DESCRIPTION
**Pull request by yogwoggf**
Summary:
- Added a new package, `luals-annotations`, which is a TypeScript project.
- Added Lest annotation generation using the docs from `@lest/docs`.
- Added new top-level NPM commands to generate an annotation and save it to a file, and also to build the project.

### CODE REVIEW

File: packages/luals-annotations/jest.config.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/** @type {import('ts-jest').JestConfigWithTsJest} */
+module.exports = {
+	preset: "ts-jest",
+	testEnvironment: "node",
+	testMatch: ["**/*.test.ts"],
+	moduleNameMapper: {
+		"@/(.*)": "<rootDir>/src/$1",
```
Comment (yogwoggf) [inline packages/luals-annotations/jest.config.js]: Will remove this, fragment from the old paths alias

**Comment by yogwoggf** [comment]
Rebased and fixed commit msgs.

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,4 @@
 packages/**/dist
 node_modules
-.idea/discord.xml
+.idea
```
Comment (Derpius) [inline .gitignore]: We should not ignore all idea config files, just those that are not of any benefit to other devs...

File: packages/luals-annotations/jest.config.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+	testEnvironment: "node",
+	testMatch: ["**/*.test.ts"],
+	collectCoverage: true,
+	collectCoverageFrom: ["src/**/{!(index),}.ts"],
+	coverageThreshold: {
+		global: {
+			lines: 85,
+			branches: 75,
+			functions: 90,
+			statements: 85,
```
Comment (Derpius) [inline packages/luals-annotations/jest.config.js]: Is there a good reason why we can't shoot for 100 here?

File: packages/luals-annotations/package.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "@lest/luals-annotations",
+  "version": "1.0.0",
```
Comment (Derpius) [inline packages/luals-annotations/package.json]: Don't think this needs to be versioned

File: packages/luals-annotations/src/luaLSType.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import { Docs } from "./json-docs";
+
+function convertFunctionProperty(property: Docs.FunctionProperty): string {
+	if (!property.parameters && !property.returns) {
+		return "fun()";
+	}
```
Comment (Derpius) [inline packages/luals-annotations/src/luaLSType.ts]: Is this optimisation needed given it adds an additional source of truth?

File: packages/luals-annotations/src/luaLSType.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+	const params = property.parameters ?? [];
+	const returns = property.returns ?? [];
+
+	const paramList = params.map((param) => `${param.name}${(param.optional && "?") || ""}:${luaLSType(param)}`);
+	const returnList = returns.map((ret) => `${luaLSType(ret)}`);
```
Comment (Derpius) [inline packages/luals-annotations/src/luaLSType.ts]: I'm not sure this format string is necessary

File: packages/luals-annotations/src/luaLSType.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	switch (docType.type) {
+		case "function":
+			return convertFunctionProperty(docType as Docs.FunctionProperty);
+		case "array":
+			return convertArrayProperty(docType as Docs.ArrayProperty);
+		case "table":
+			return convertTableProperty(docType as Docs.TableProperty);
+		default:
+			return docType.type;
+	}
```
Comment (Derpius) [inline packages/luals-annotations/src/luaLSType.ts]: Couple of alternative options to avoid having to cast the type, although not a blocker:
- Make the property type a generic
- Use the same approach as the site with `isBlahProperty` functions that assert the type for you

File: packages/luals-annotations/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+const document = new AnnotationBuilder();
+document.add("---@meta");
+
+classes.forEach((classDef) => {
+	document.withClass(classDef.name, classDef.description, false, () => {
+		classDef.fields?.forEach((field) => document.addField(field));
+		document.addClassDeclaration();
+		classDef.methods?.forEach((method) => document.addFunction(method));
+	});
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts]: I feel like this would benefit from changing `withClass` to take an object of options, as I have no idea what `false` indicates here

File: packages/luals-annotations/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	document.addClassDeclaration();
+	matchers.forEach((matcher) => document.addFunction(matcher));
+});
+
+document.withClass("lest.InverseMatchers", "Inverse matchers", true, () => {
+	document.addClassDeclaration();
+	matchers.forEach((matcher) => document.addFunction(matcher));
+});
+
+if (process.env["npm_config_debug"]) {
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts:68]: What's this?

File: packages/luals-annotations/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import { classes, functions, matchers } from "./json-docs";
+import AnnotationBuilder from "./annotationBuilder";
+import * as fs from "fs";
+
+const document = new AnnotationBuilder();
+document.add("---@meta");
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts]: I feel like this should be added by default as part of `AnnotationBuilder`

File: packages/luals-annotations/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	process.exit(0);
+}
+
+const targetFilePath = process.argv[2];
+if (!targetFilePath) {
+	console.error("No target file path provided");
+	process.exit(1);
+}
+
+fs.writeFileSync(targetFilePath, document.toString());
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts]: I don't really like `toString` here. It looks like it just calls `.build()` so can we not call that here instead?

File: packages/luals-annotations/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import * as fs from "fs";
+
+const document = new AnnotationBuilder();
+document.add("---@meta");
+
+classes.forEach((classDef) => {
+	document.withClass(classDef.name, classDef.description, false, () => {
+		classDef.fields?.forEach((field) => document.addField(field));
+		document.addClassDeclaration();
+		classDef.methods?.forEach((method) => document.addFunction(method));
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts]: I removed the `methods` property before merging the docs. All functions should just be part of `fields`,

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+			throw new Error("Cannot add class declaration without a class");
+		}
+
+		this.add(`${this.currentClass} = {}`);
+	}
+
+	addField(property: Docs.Property) {
+		if (!this.currentClass) {
+			throw new Error("Cannot add field without a class");
+		}
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts]: Given these need to be called from within `withClass`, could they just be non-class functions that are passed to `classFn`?

This way they're only exposed in the context where they can be called.

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+		let description = property.description;
+		if (typeof description !== "string") {
+			// Unfortunately, we can't really have multi-line descriptions for fields
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts]: Might be one to look into for a workaround, or one to raise on the LuaLS github

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+			description = description.join(" ");
+		}
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts]: I'm not sure we need this `let` at all (we should generally use `const` 99.9% of the time). Is this logic not simple enough to be a ternary?

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+		);
+		const returnList = returns.map((ret) => `---@return ${luaLSType(ret)} ${ret.name ?? ""}`);
+
+		const functionCharacter = this.staticMethod ? "." : ":";
+		const functionPrefix = this.currentClass ? `${this.currentClass}${functionCharacter}` : "";
+
+		const renderFunction = (signature: string) => {
+			this.addDescription(func.description);
+			this.add(...paramList, ...returnList, signature);
+		};
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts]: We should generally avoid defining functions inside other functions, except when writing a higher order function.

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+		if (func.aliases) {
+			func.aliases.forEach((alias) => {
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts]: Can we use optional chaining (`?.`) here?

**Review by Derpius — CHANGES_REQUESTED** [review]
Only covered things that immediately stood out or are quite serious to try and help keep the review process efficient.

I'll do another pass once the current comments have been resolved.

Note that I haven't taken a look at the tests yet.

File: packages/luals-annotations/src/index.ts (hunk 9 above)
Comment (yogwoggf) [reply packages/luals-annotations/src/index.ts:68]: Forgot to comment that, but npm has a feature where any double dash argument (like `--debug`) gets passed onto the env as npm_config_X.

File: packages/luals-annotations/jest.config.js (hunk 3 above)
Comment (yogwoggf) [reply packages/luals-annotations/jest.config.js]: I DONT KNOW EITHER!!!!!!
Jest says that the luaLSType file is like at 98% branches. I checked the line number it was complaining for and it's just an empty line.
I also set it low because the annotation builder tests weren't done yet. Should be fine shooting for 95-100 now.

Comment (yogwoggf) [reply packages/luals-annotations/jest.config.js]: Done

File: packages/luals-annotations/src/annotationBuilder.ts (hunk 14 above)
Comment (yogwoggf) [reply packages/luals-annotations/src/annotationBuilder.ts]: There's not exactly a workaround here, it's just that fields are primitive annotations which means that they're parsed as a single line. Could raise this as an issue on the LuaLS github later.

File: .gitignore (hunk 2 above)
Comment (yogwoggf) [reply .gitignore]: Removed vcs.xml and workspace.xml because they're also not important
- vcs.xml: auto-generated upon launch
- workspace.xml: store things like last opened file

File: packages/docs/package.json [hunk 18]
```diff
@@ -4,6 +4,12 @@
   "description": "API documentation in JSON form.",
   "main": "dist/index",
   "types": "dist/index",
+  "exports": {
+    ".": "./dist/index.js",
+    "./functions": "./dist/functions/index.js",
+    "./types": "./dist/types/index.js",
+    "./matchers": "./dist/matchers/index.js"
+  },
```
Comment (yogwoggf) [inline packages/docs/package.json:12]: The annotations requires this because it categorizes the docs by importing the individual folders instead of getting them through the main entrypoint. Also, it seems to require file extensions and only works with the `NodeNext` module resolution.

File: .gitignore (hunk 2 above)
Comment (Derpius) [reply .gitignore]: `vcs.xml` should not be removed. We don't use it for any specific settings at the moment, but it can contain important configuration options.

`workspace.xml` is already ignored by `.idea/.gitignore`, which Jetbrains editors add by default.

For an exhaustive list of everything that should be ignored by ALL Jetbrains editors (i.e. we do not need anywhere near this many ignored files/folders), see this list:
https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore

File: packages/luals-annotations/src/annotationBuilder.ts (hunk 14 above)
Comment (Derpius) [reply packages/luals-annotations/src/annotationBuilder.ts]: We don't need a workaround even, it's 100% supported by LuaLS

https://github.com/LuaLS/lua-language-server/issues/1531#issuecomment-1239129419

**Comment by Derpius** [comment]
There's a couple things left that need resolving, after which I'll do a second pass:

- https://github.com/TAServers/lest/pull/46#discussion_r1257214571
- https://github.com/TAServers/lest/pull/46#discussion_r1257217842

There are a couple bits of logic I saw while resolving threads that are a little hard to parse, so you might want to do a quick self-review of your changes in case there's anything that you think could be improved still

File: .gitignore (hunk 2 above)
Comment (yogwoggf) [reply .gitignore]: As discussed, turns out this was my issue. All idea config files are corrected.

File: packages/luals-annotations/src/annotationBuilder.ts (hunk 14 above)
Comment (yogwoggf) [reply packages/luals-annotations/src/annotationBuilder.ts]: Don't know how I missed that?? :trollface:

**Comment by yogwoggf** [comment]
FYI: In the second pass I assume you will look at the tests. Because the generated annotations aren't strict about whitespace/formatting, the tests look for "expected elements" and "unexpected elements." These effectively make sure that each annotation is a certain shape without considering whitespace or formatting.

File: package.json [hunk 19]
```diff
@@ -9,6 +9,9 @@
     "format:site": "prettier --check packages/site",
     "format:docs": "prettier --check packages/docs",
     "build:lest": "npm run build --workspace=packages/lest",
+    "build:luals-annotations": "npm run build:docs && npm run build --workspace=packages/luals-annotations",
+    "test:luals-annotations": "npm run test --workspace=packages/luals-annotations",
+    "annotate:luals": "npm run start --workspace=packages/luals-annotations",
```
Comment (Derpius) [inline package.json:14]: nit: Inconsistency between `luals-annotations` and `luals`. Also, could you group them like the other scripts?

File: packages/luals-annotations/jest.config.js
Comment (Derpius) [inline packages/luals-annotations/jest.config.js:1]: Not a review comment, but you can actually make jest configs with typescript if you have `ts-node` as a dependency

File: packages/luals-annotations/jest.config.js [hunk 20]
```diff
@@ -0,0 +1,16 @@
+/** @type {import('ts-jest').JestConfigWithTsJest} */
+module.exports = {
+	preset: "ts-jest",
+	testEnvironment: "node",
+	testMatch: ["**/*.test.ts"],
+	collectCoverage: true,
+	collectCoverageFrom: ["src/**/{!(index),}.ts"],
```
Comment (Derpius) [inline packages/luals-annotations/jest.config.js:7]: Is there a reason we're excluding coverage for index files? Surely the tests should import from the indexes at least once?

File: packages/luals-annotations/package.json [hunk 21]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "@lest/luals-annotations",
+  "description": "Package to automatically create LuaLS annotations for Lest",
+  "main": "dist/index",
+  "scripts": {
+    "start:ts-node": "ts-node src/index.ts",
+    "start": "node dist/src/index.js",
+    "dev": "npm run --debug start:ts-node && onchange \"src/**/*.ts\" -- npm run --debug start:ts-node",
```
Comment (Derpius) [inline packages/luals-annotations/package.json:8]: nodemon works out the box with ts-node, but I've actually been using `tsc-watch` recently as `ts-node` has numerous compatibility issues (especially in a monorepo config)

Can leave this as is, but I recommend nodemon or tsc-watch

File: packages/luals-annotations/package.json [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+    "build": "tsc -p tsconfig.build.json",
```
Comment (Derpius) [inline packages/luals-annotations/package.json:9]: Should this set `--build`?

File: packages/luals-annotations/src/index.ts [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+
+function generateClass(classInfo: Class) {
+	const { name, description, fields = [] } = classInfo;
+	const cls = new ClassBuilder({
+		name,
+		description,
+	});
+
+	const classMethods = fields.filter((method) => isFunctionProperty(method)) ?? [];
+	const classFields = fields.filter((property) => !isFunctionProperty(property)) ?? [];
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts:23]: I don't think `Array.prototype.filter` ever returns null or undefined, are these nullish coalescing operators doing anything?

File: packages/luals-annotations/src/index.ts [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+import AnnotationBuilder, { ClassBuilder } from "./annotationBuilder";
+
+const functionDocs = Object.values<Function>(functions);
+const classDocs = Object.values<Class>(classes);
+const matcherDocs = Object.values<Function>(matchers);
+
+const document = new AnnotationBuilder();
+
+function generateClass(classInfo: Class) {
+	const { name, description, fields = [] } = classInfo;
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts:16]: nit: Can you destructure this in the function definition above?

File: packages/luals-annotations/src/index.ts [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+		name,
+		description,
+	});
+
+	const classMethods = fields.filter((method) => isFunctionProperty(method)) ?? [];
+	const classFields = fields.filter((property) => !isFunctionProperty(property)) ?? [];
+
+	classFields.forEach((field) => cls.addField(field));
+	cls.addDeclaration();
+	classMethods.forEach((method) => cls.addFunction(method as Function));
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts:27]: This cast is almost certainly wrong, `Function` and `FunctionProperty` are _not_ the same interface

File: packages/luals-annotations/src/index.ts [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+	document.addClass(cls);
+}
+
+function generateClasses() {
+	classDocs.forEach(generateClass);
+}
+
+function generateFunctions() {
+	functionDocs.forEach((func) => document.addFunction(func));
+}
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts:37]: I don't like that these functions are _exclusively_ side effects. I think we can just drop these functions entirely given they're so simple?

File: packages/luals-annotations/src/index.ts [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+	document.addClass(matchersClass);
+}
+
+generateClasses();
+generateFunctions();
+generateMatcherClass({ inverse: false });
+generateMatcherClass({ inverse: true });
+
+// This environment variable is set by NPM when you pass a double dashed argument to it.
+// The dev script passes --debug to the start:ts-node script, so this code is ran when you run `npm run dev`.
```
Comment (Derpius) [inline packages/luals-annotations/src/index.ts:67]: nit: This comment will get out of date _very_ quickly given it's coupled to the package config

I think we can replace both of these lines with just:
```ts
// This environment variable is set by NPM when you pass --debug to it
```

File: packages/luals-annotations/src/luaLSType.ts [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+	ArrayProperty,
+	TableProperty,
+	isArrayProperty,
+	isFunctionProperty,
+	isTableProperty,
+	Property,
+} from "@lest/docs";
+
+export function formatProperty(param: Property, typeSeparator = ":"): string {
+	const { name, optional } = param;
```
Comment (Derpius) [inline packages/luals-annotations/src/luaLSType.ts:12]: nit: Can we destructure this in the function signature?

File: packages/luals-annotations/src/luaLSType.ts [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+	FunctionProperty,
+	ArrayProperty,
+	TableProperty,
+	isArrayProperty,
+	isFunctionProperty,
+	isTableProperty,
+	Property,
+} from "@lest/docs";
+
+export function formatProperty(param: Property, typeSeparator = ":"): string {
```
Comment (Derpius) [inline packages/luals-annotations/src/luaLSType.ts:11]: Exposing this as `typeSeparator` seems like too much control given to the callers (it needs to know about the internals of this function)

It looks like you only use this to do `formatProperty(param, " ")` in another file. You should just not export this function and replace that code. This function should have a [single responsibility](https://www.digitalocean.com/community/conceptual-articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#single-responsibility-principle)

File: packages/luals-annotations/src/luaLSType.ts [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+	const { name, optional } = param;
+	return `${name}${optional ? "?" : ""}${typeSeparator}${luaLSType(param)}`;
+}
+
+function convertFunctionProperty(property: FunctionProperty): string {
+	const { parameters = [], returns = [] } = property;
```
Comment (Derpius) [inline packages/luals-annotations/src/luaLSType.ts:17]: nit: Can you destructure this in the function signature?

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+
+	const functionCharacter = staticMethod ? "." : ":";
+	const functionPrefix = className ? `${className}${functionCharacter}` : "";
+
+	const functionNames = [func.name, ...(func.aliases ?? [])];
+
+	return functionNames.map(
+		(name) => `function ${functionPrefix}${name}(${parameters.map((param) => param.name)}) end`
+	);
+}
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts:28]: This code is quite difficult to read. A lot of the variables don't make sense, or don't need to exist, and some things that aren't variables (like mapping the parameters) should be.

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 32]
```diff
@@ -0,0 +1,127 @@
...
+function renderParameterAnnotations(parameters: Property[]): string[] {
+	return parameters.map((param) => `---@param ${formatProperty(param, " ")}`);
+}
+
+function renderReturnAnnotations(returns: Property[]): string[] {
+	return returns.map((ret) => `---@return ${luaLSType(ret)} ${ret.name ?? ""}`);
+}
+
+function renderFunctionAliases(func: Function, staticMethod: boolean, className: string): string[] {
+	const { parameters = [] } = func;
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts:18]: nit: Can you destructure this in the function signature? Also, can you replace `func.blah` below with destructuring for consistency's sake

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 33]
```diff
@@ -0,0 +1,127 @@
...
+function renderDescription(description: string | string[]): string[] {
+	const descriptionLines = Array.isArray(description) ? description : [description];
+	return descriptionLines.map((line) => `--- ${line}`);
+}
+
+function renderFunctionDeclaration(
+	func: Function,
+	{ staticMethod = false, className = "" }: FunctionRenderOptions = {}
+): string {
+	const { parameters = [], returns = [], description = [] } = func;
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts:39]: nit: Can you destructure this in the function signature?

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 34]
```diff
@@ -0,0 +1,127 @@
...
+
+	// LuaLS has a very strict order about the function annotations, so this is why we have a specific order.
+	// 1. Description
+	// 2. Parameters
+	// 3. Returns
+	// 4. Function signature
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts:45]: nit: This comment is mostly useless (the order is clear from the code). I'd argue that the comment in it's entirety isn't really necessary either

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 35]
```diff
@@ -0,0 +1,127 @@
...
+		this.lines.push(line);
+	}
+
+	build(): string {
+		return this.lines.join("\n");
+	}
+
+	buildToLines(): string[] {
+		return this.lines;
+	}
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts:69]: What's the use of this function? Can you not just access `lines`?

File: packages/luals-annotations/src/annotationBuilder.ts [hunk 36]
```diff
@@ -0,0 +1,127 @@
...
+		super();
+		this.add("---@meta");
+	}
+
+	addFunction(func: Function) {
+		this.add(renderFunctionDeclaration(func));
+	}
+
+	addClass(cls: ClassBuilder) {
+		cls.buildToLines().forEach((line) => this.add(line));
```
Comment (Derpius) [inline packages/luals-annotations/src/annotationBuilder.ts:125]: This seems really unnecessary when we can just do `this.lines = [...this.lines, ...cls.lines]`, or `this.push(...cls.lines)` or any concat method

**Review by Derpius — COMMENTED** [review]
Still haven't reviewed the tests yet, as there were still quite a few issues

File: packages/luals-annotations/src/annotationBuilder.ts (hunk 35 above)
Comment (yogwoggf) [reply packages/luals-annotations/src/annotationBuilder.ts:69]: Change from a while ago, but the interface was already being depended upon so.. I kind of half-assed it.

File: packages/luals-annotations/package.json (hunk 21 above)
Comment (yogwoggf) [reply packages/luals-annotations/package.json:8]: Yup, ts-node is awful with monorepos, will look into a switch (nodemon/tsc-watch)

File: package.json (hunk 19 above)
Comment (yogwoggf) [reply package.json:14]: This was in the acceptance criteria, but I'll change it.

File: packages/luals-annotations/jest.config.js (hunk 20 above)
Comment (yogwoggf) [reply packages/luals-annotations/jest.config.js:7]: Again, another issue caused by a change when I had multiple index.ts files. `index.ts` can be considered a unit, but all it does is just combine all of the other **tested** units together to create the annotation file, it doesn't export any functions. It's pretty much the application.

File: packages/luals-annotations/src/index.ts (hunk 24 above)
Comment (yogwoggf) [reply packages/luals-annotations/src/index.ts:16]: No

**Comment by yogwoggf** [comment]
Still planning to fix these markups...

**Comment by Derpius** [comment]
@yogwoggf when are you finishing this?

**Comment by yogwoggf** [comment]
> @yogwoggf when are you finishing this?

I'm not sure but it's 100% on the farthest backburner since Lest has had no activity since July 16th.
